ui/StatusQ/src/statusemojimodel.cpp

```diff
@@ -0,0 +1,227 @@
+                                 m_emojiJson.cend(),
+                                 [shortname](const auto &emoji) {
+                                     return emoji.isObject()
+                                            && emoji.toObject().value(kShortname).toString()
+                                                   == shortname;
+                                 });
+    if (it != m_emojiJson.cend())
+        return it->toObject().value(kUnicode).toString();
+    return {};
+}
+
+int StatusEmojiModel::getCategoryOffset(int categoryIndex) const {
+    if (categoryIndex <= 0 || categoryIndex >= categories().size())
+        return 0;
+
+    const auto categoryName = categories().at(categoryIndex);
+    const auto catIt = std::find_if(m_emojiJson.cbegin(),
+                                    m_emojiJson.cend(),
+                                    [categoryName](const auto &emoji) {
+                                        return emoji.isObject()
+                                        && emoji.toObject().value(kCategory).toString()
+                                            == categoryName;
+                                    });
+    if (catIt != m_emojiJson.cend())
+        return std::distance(m_emojiJson.cbegin(), catIt) + m_recentEmojiJson.size();
+    return 0;
+}
+
+QStringList StatusEmojiModel::categories() const
+{
+    static const QStringList categories{kRecentCategoryName,
+                                        QStringLiteral("smileys, people & body"),
+                                        QStringLiteral("animals & nature"),
+                                        QStringLiteral("food & drink"),
+                                        QStringLiteral("travel & places"),
+                                        QStringLiteral("activities"),
+                                        QStringLiteral("objects"),
+                                        QStringLiteral("symbols"),
+                                        QStringLiteral("flags")};
+    return categories;
+}
+
+QStringList StatusEmojiModel::recentEmojis() const
+{
+    return m_recentEmojis;
+}
+
+void StatusEmojiModel::setRecentEmojis(const QStringList &newRecentEmojis)
+{
+    if (m_recentEmojis == newRecentEmojis)
+        return;
+    m_recentEmojis = newRecentEmojis;
+    cleanAndResizeRecentEmojis();
+    addRecentEmojisToModel(m_recentEmojis);
+    emit recentEmojisChanged();
+}
+
+void StatusEmojiModel::addRecentEmoji(const QString &hexcode)
+{
+    m_recentEmojis.prepend(hexcode);
+    cleanAndResizeRecentEmojis();
+    addRecentEmojisToModel(m_recentEmojis);
+    emit recentEmojisChanged();
+}
+
+void StatusEmojiModel::cleanAndResizeRecentEmojis()
+{
+    m_recentEmojis.removeDuplicates();
+    if (m_recentEmojis.size() > MAX_EMOJI_NUMBER) {
+        while (m_recentEmojis.size() > MAX_EMOJI_NUMBER)
+            m_recentEmojis.removeLast();
+    }
+}
+
+void StatusEmojiModel::addRecentEmojisToModel(const QStringList &emojiHexcodes)
```

Comment:
Do we still save the recent emojis to settings or something so that they are available even after an app reset

Reply:
Yeah we still do from NIM using `localAppSettings`. This here is to add them to the model so that they can be served together with the regular ones from one place.